File: app/src/main/java/net/emilla/action/box/SnippetsFragment.java
package net.emilla.action.box;

import static net.emilla.chime.Chime.PEND;
import static net.emilla.chime.Chime.RESUME;

import android.content.SharedPreferences;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.Nullable;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import net.emilla.R;
import net.emilla.activity.AssistActivity;
import net.emilla.annotation.internal;
import net.emilla.databinding.FragmentItemListBinding;
import net.emilla.sort.ItemSearchAdapter;
import net.emilla.util.Dialogs;

public final class SnippetsFragment extends ActionBox {

    @internal SnippetsFragment() {
        super(R.layout.fragment_item_list);
    }

    public static SnippetsFragment newInstance() {
        return new SnippetsFragment();
    }

    private /*late*/ SharedPreferences mPrefs;

    private ItemSearchAdapter<Snippet> mAdapter;

    @Override
    public View onCreateView(
        LayoutInflater inflater,
        @Nullable ViewGroup container,
        @Nullable Bundle savedInstanceState
    ) {
        var binding = FragmentItemListBinding.inflate(inflater, container, false);
        return binding.recycler;
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        var act = (AssistActivity) requireActivity();
        var recycler = (RecyclerView) view;

        var manager = new LinearLayoutManager(act);
        manager.setReverseLayout(true);

        recycler.setLayoutManager(manager);

        var inflater = act.getLayoutInflater();
        mPrefs = act.getSharedPreferences();
        mAdapter = Snippet.adapter(
            inflater,
            mPrefs,
            snippet -> giveText(act, snippet.displayName, snippet.text(mPrefs))
        );
        recycler.setAdapter(mAdapter);
    }

    @Nullable
    private Snippet selectedSnippet(AssistActivity act, String search) {
        Snippet selected = mAdapter.preferredItem(search);
        if (selected == null) {
            act.chime(PEND);
        }
        return selected;
    }

    public void peek(String snippetLabel) {
        var act = (AssistActivity) requireActivity();

        Snippet snippet = selectedSnippet(act, snippetLabel);
        if (snippet != null) {
            giveText(act, snippet.displayName, snippet.text(mPrefs));
        }
    }

    public void copy(String snippetLabel) {
        var act = (AssistActivity) requireActivity();

        Snippet snippet = selectedSnippet(act, snippetLabel);
        if (snippet != null) {
            giveCopy(act, snippet.text(mPrefs));
        }
    }

    public void pop(String snippetLabel) {
        var act = (AssistActivity) requireActivity();

        Snippet snippet = selectedSnippet(act, snippetLabel);
        if (snippet != null) {
            snippet.delete(act, mPrefs);
            mAdapter.remove(snippet);
            giveCopy(act, snippet.text(mPrefs));
        }
    }

    public void remove(String snippetLabel) {
        var act = (AssistActivity) requireActivity();

        Snippet snippet = selectedSnippet(act, snippetLabel);
        if (snippet != null) {
            snippet.delete(act, mPrefs);
            mAdapter.remove(snippet);
            act.give(a -> {});
        }
    }

    public void add(String snippetLabel, String text) {
        var act = (AssistActivity) requireActivity();

        var snippet = new Snippet(snippetLabel);
        if (mAdapter.exactItem(snippetLabel) != null) {
            var res = act.getResources();
            offerDialog(
                act, Dialogs.dual(
                    act, R.string.dialog_overwrite_snippet,

                    res.getString(R.string.dlg_msg_overwrite_snippet, snippetLabel),
                    R.string.overwrite,
                    (dlg, which) -> {
                        snippet.overwrite(act, mPrefs, text);
                        act.chime(RESUME);
                    }
                )
            );
        } else {
            snippet.saveNew(act, mPrefs, text);
            mAdapter.add(snippet);
            act.give(a -> {});
        }
    }

    @Override
    public void instruct(@Nullable String instruction) {
        if (mAdapter != null) {
            // TODO: figure out how this damn life cycle works. The adapter should never be null
            //  while the fragment is alive, but this method may be called outside of that window.
            //  I want to actually understand when and how this happens and how to generally avoid
            //  it.
            mAdapter.search(instruction);
        }
    }

}

File: app/src/main/java/net/emilla/action/box/SnippetsFragment.kt
package net.emilla.action.box

import android.content.DialogInterface
import android.os.Bundle
import android.view.View
import androidx.annotation.StringRes
import androidx.fragment.app.viewModels
import androidx.preference.PreferenceManager
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import net.emilla.R
import net.emilla.chime.Chime.PEND
import net.emilla.chime.Chime.RESUME
import net.emilla.command.core.SnippetAction
import net.emilla.config.SettingVals
import net.emilla.util.Dialogs

class SnippetsFragment : ActionBox(R.layout.snippet_item_list) {

    companion object {
        @JvmStatic
        fun newInstance() = SnippetsFragment()
    }

    private val vm: SnippetsViewModel by viewModels<SnippetsViewModel> {
        val prefs = PreferenceManager.getDefaultSharedPreferences(activity)
        SnippetsViewModel.Factory(prefs)
    }

    private val adapter: SnippetAdapter by lazy {
        SnippetAdapter(vm.snippetLabels(), SnippetAdapter.OnItemClickListener { pos ->
            peek(vm.labelAt(pos))
        })
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        view as RecyclerView

        val mgr = LinearLayoutManager(activity)
        mgr.setStackFromEnd(true)

        view.setLayoutManager(mgr)
        view.setAdapter(adapter)
    }

    fun contains(label: String) = vm.snippetLabels().contains(label)

    val isEmpty get() = vm.snippetLabels().isEmpty

    fun prime(action: SnippetAction) {
        activity.chime(PEND)

        adapter.setItemClickAction(
        when (action) {
            SnippetAction.PEEK -> { pos ->
                peek(vm.labelAt(pos))
            }
            SnippetAction.GET -> { pos ->
                val label = vm.labelAt(pos)
                get(label)
                resetItemClickListener()
            }
            SnippetAction.POP -> { pos ->
                val label = vm.labelAt(pos)
                pop(label, label)
                resetItemClickListener()
            }
            SnippetAction.REMOVE -> { pos ->
                val label = vm.labelAt(pos)
                remove(label, label)
                give {}
                resetItemClickListener()
            }
            SnippetAction.ADD -> { pos ->
                val label = vm.labelAt(pos)
                val text = activity.dataText()
                if (text != null) tryAdd(label, text)
                else peek(label)
                resetItemClickListener()
            }
        })
    }

    private fun resetItemClickListener() {
        adapter.setItemClickAction { pos ->
            peek(vm.labelAt(pos))
        }
    }

    fun peek(label: String) = giveText(SettingVals.snippet(vm.prefs, label))

    fun get(label: String) = giveCopy(SettingVals.snippet(vm.prefs, label))

    fun pop(label: String, lcLabel: String) {
        get(lcLabel)
        remove(label, lcLabel)
    }

    fun tryAdd(label: String, text: String) {
        if (contains(label)) {
            val msg = str(R.string.dlg_msg_overwrite_snippet, label)
            offerDialog(
                Dialogs.dual(activity, R.string.dialog_overwrite_snippet, msg, R.string.overwrite,
                DialogInterface.OnClickListener { dlg, which ->
                    vm.replaceSnippet(label, text, adapter)
                    chime(RESUME)
                })
            )
        } else {
            vm.addSnippet(label, text, adapter)
            give { toast(R.string.toast_saved) }
        }
    }

    fun remove(label: String, lcLabel: String) {
        vm.removeSnippet(lcLabel, adapter)
        toast(R.string.toast_snippet_deleted, label)
    }

    @StringRes
    override val name = R.string.command_snippets
}
